cleverhans/attacks.py

```diff
@@ -48,48 +65,98 @@ def generate(self, x, **kwargs):
-        that is not fully expressed symbolically.
+        Generate adversarial examples and return them as a Numpy array.
+        Sub-classes *should not* implement this method unless they must
+        perform special handling of arguments.
         :param x_val: A Numpy array with the original inputs.
         :param **kwargs: optional parameters used by child classes.
         :return: A Numpy array holding the adversarial examples.
         """
         if self.back == 'th':
             raise NotImplementedError('Theano version not implemented.')
 
-        if not self.inf_loop:
-            self.inf_loop = True
-            import tensorflow as tf
-
-            # Generate this attack's graph if not done previously
-            if not hasattr(self, "_x") and not hasattr(self, "_x_adv"):
-                input_shape = list(x_val.shape)
-                input_shape[0] = None
-                self._x = tf.placeholder(tf.float32, shape=input_shape)
-                self._x_adv = self.generate(self._x, **kwargs)
-            self.inf_loop = False
-        else:
-            error = "No symbolic or numeric implementation of attack."
-            raise NotImplementedError(error)
+        import tensorflow as tf
 
         if self.sess is None:
             raise ValueError("Cannot use `generate_np` when no `sess` was"
                              " provided")
-        return self.sess.run(self._x_adv, feed_dict={self._x: x_val})
+
+        # the set of arguments that are structural properties of the attack
+        # if these arguments are different, we must construct a new graph
+        fixed = dict((k, v) for k, v in kwargs.items()
+                     if k in self.structural_kwargs)
+
+        # the set of arguments that are passed as placeholders to the graph
+        # on each call, and can change without constructing a new graph
+        feedable = dict((k, v) for k, v in kwargs.items()
+                        if k in self.feedable_kwargs)
+
+        if len(fixed) + len(feedable) < len(kwargs):
+            warnings.warn("Supplied extra keyword arguments that are not "
+                          "used in the graph computation. They have been "
+                          "ignored.")
+
+        if not all(isinstance(value, collections.Hashable)
+                   for value in fixed.values()):
+            # we have received a fixed value that isn't hashable
+            # this means we can't cache this graph for later use,
+            # and it will have to be discarded later
+            hash_key = None
+        else:
+            # create a unique key for this set of fixed paramaters
+            hash_key = tuple(sorted(fixed.items()))
+
+        if hash_key not in self.graphs:
+            # try our very best to create a TF placeholder for each of the
+            # feedable keyword arguments, and check the types are one of
+            # the allowed types
```

Comment:
Not necessary to merge, but consider putting this list of types in utils.py where it could be used by other functions